src/lem/constrainer.rs

```diff
@@ -0,0 +1,344 @@
+        enforce_selector_with_premise, implies_equal, implies_equal_zero,
+    },
+    pointer::AllocatedPtr,
+};
+
+use crate::field::{FWrap, LurkField};
+
+use super::{pointers::ZPtr, store::Store, Witness, LEM, LEMOP};
+
+/// Manages allocations for numeric variables in a constraint system
+#[derive(Default)]
+struct AllocationManager<F: LurkField>(HashMap<FWrap<F>, AllocatedNum<F>>);
+
+impl<F: LurkField> AllocationManager<F> {
+    /// Checks if the allocation for a numeric variable has already been cached.
+    /// If so, return the cached allocation variable. Allocate, cache and return
+    /// otherwise.
+    pub(crate) fn alloc<CS: ConstraintSystem<F>>(
+        &mut self,
+        cs: &mut CS,
+        f: F,
+    ) -> Result<AllocatedNum<F>, String> {
+        let wrap = FWrap(f);
+        match self.0.get(&wrap) {
+            Some(alloc) => Ok(alloc.to_owned()),
+            None => {
+                let digits = f.hex_digits();
+                let alloc =
+                    AllocatedNum::alloc(cs.namespace(|| format!("allocate {digits}")), || Ok(f))
+                        .map_err(|e| format!("{e}: couldn't allocate {digits}"))?;
+                self.0.insert(wrap, alloc.clone());
+                Ok(alloc)
+            }
+        }
+    }
+}
+
+impl LEM {
+    fn allocate_ptr<F: LurkField, CS: ConstraintSystem<F>>(
+        cs: &mut CS,
+        z_ptr: &ZPtr<F>,
+        name: &String,
+    ) -> Result<AllocatedPtr<F>, String> {
+        let alloc_tag =
+            AllocatedNum::alloc(cs.namespace(|| format!("allocate {name}'s tag")), || {
+                Ok(z_ptr.tag.to_field())
+            })
+            .map_err(|e| format!("{e}: couldn't allocate {name}'s tag"))?;
+        let alloc_hash =
+            AllocatedNum::alloc(cs.namespace(|| format!("allocate {name}'s hash")), || {
+                Ok(z_ptr.hash)
+            })
+            .map_err(|e| format!("{e}: couldn't allocate {name}'s hash"))?;
+        Ok(AllocatedPtr::from_parts(&alloc_tag, &alloc_hash))
+    }
+
+    fn inputize_ptr<F: LurkField, CS: ConstraintSystem<F>>(
+        cs: &mut CS,
+        alloc_ptr: &AllocatedPtr<F>,
+        name: &String,
+    ) -> Result<(), String> {
+        alloc_ptr
+            .tag()
+            .inputize(cs.namespace(|| format!("inputize {}'s tag", name)))
+            .map_err(|e| format!("{}: couldn't inputize {}'s tag", &e, name))?;
+        alloc_ptr
+            .hash()
+            .inputize(cs.namespace(|| format!("inputize {}'s hash", name)))
+            .map_err(|e| format!("{}: couldn't inputize {}'s hash", &e, name))?;
+        Ok(())
+    }
+
+    fn on_concrete_path(concrete_path: &Option<Boolean>) -> Result<bool, String> {
+        match concrete_path {
+            None => Ok(true),
+            Some(concrete_path) => match concrete_path.get_value() {
+                Some(b) => Ok(b),
+                None => Err("Couldn't check whether we're on a concrete path".to_string()),
+            },
```

Comment:
```suggestion
            Some(concrete_path) => concrete_path
                .get_value()
                .ok_or_else(|| anyhow!("Couldn't check whether we're on a concrete path")),,
```

Reply:
Can't apply suggestion due to conflict with previous suggestion